File: BaekJoon/2822_점수계산_Sorting.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

// 점수계산 2822 Sorting
// O( 5 log5 + 3 ) = O(1)


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    // {점수, 번호}
    vector<pair<int, int>> scores(8);
    for (int i = 0; i < 8; ++i)
    {
        int x;
        cin >> x;
        scores[i] = { x, i + 1 };   // 번호는 1-based indexing
    }

    // 상위 5개만 오름차순 부분 정렬
    partial_sort(scores.begin(), scores.begin() + 5, scores.end(),
        [](const auto& a, const auto& b) {return a.first > b.first; });

    int sum{ 0 };       // 총합계
    vector<int> idx(5); // 상위 5개 인덱스

    for (int i = 0; i < 5; ++i)
    {
        sum += scores[i].first;
        idx[i] = scores[i].second;
    }

    sort(idx.begin(), idx.end());

    cout << sum << "\n";
    for (int i : idx) cout << i << " ";

    
    return 0;
}

File: BaekJoon/Java/2822_점수계산_Sorting.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

// 점수계산 2822 Sorting
// O( 5 log5 + 3 ) = O(1)


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[][] scores = new int[8][2];
        for (int i = 0; i < 8; ++i) {
            scores[i][0] = Integer.parseInt(br.readLine());
            scores[i][1] = i + 1;
        }

        // 점수 기준 내림차순
        Arrays.sort(scores, (a, b) -> Integer.compare(b[0], a[0]));

        int sum = 0;
        int[] idx = new int[5];

        for (int i = 0; i < 5; ++i) {
            sum += scores[i][0];
            idx[i] = scores[i][1];
        }

        Arrays.sort(idx);

        System.out.println(sum);
        for (int i : idx) {
            System.out.print(i + " ");
        }
    }
}
